core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1543,15 +1543,19 @@ class KafkaApis(val requestChannel: RequestChannel,
     var skippedMarkers = 0
     for (marker <- markers.asScala) {
       val producerId = marker.producerId
-      val (goodPartitions, partitionsWithIncorrectMessageFormat) = marker.partitions.asScala.partition { partition =>
+
+      val (goodPartitions, unknownOrIncorrectMessageFormatPartitions) = marker.partitions.asScala.partition { partition =>
         replicaManager.getMagic(partition) match {
           case Some(magic) if magic >= RecordBatch.MAGIC_VALUE_V2 => true
           case _ => false
         }
       }
 
-      if (partitionsWithIncorrectMessageFormat.nonEmpty) {
+      val (partitionsWithIncorrectMessageFormat, unknownPartitions) = unknownOrIncorrectMessageFormatPartitions.partition(metadataCache.contains(_))
+
+      if (unknownOrIncorrectMessageFormatPartitions.nonEmpty) {
         val currentErrors = new ConcurrentHashMap[TopicPartition, Errors]()
+        unknownPartitions.foreach { partition => currentErrors.put(partition, Errors.UNKNOWN_TOPIC_OR_PARTITION)}
         partitionsWithIncorrectMessageFormat.foreach { partition => currentErrors.put(partition, Errors.UNSUPPORTED_FOR_MESSAGE_FORMAT) }
```

Comment:
Discussed offline, but I think we need to ensure that we are only returning this case when we _know_ that the message format version is lower than 2. That means we probably need to handle the case that the partition is in the metadata cache, but we have no local replica (i.e. we have not yet received the LeaderIsr).